sdks/python/apache_beam/runners/common.py

```diff
@@ -51,117 +51,150 @@ def receive(self, windowed_value):
+    self.finish_bundle_method = None
+    self.do_fn = None
+
+  @staticmethod
+  def create_signature(do_fn):
+    """Creates the signature fo a given DoFn object."""
+    assert isinstance(do_fn, core.DoFn)
+    signature = DoFnSignature()
+
+    def _create_do_fn_method(do_fn, method_name):
+      arguments, _, _, defaults = do_fn.get_function_arguments(method_name)
+      defaults = defaults if defaults else []
+      method_value = getattr(do_fn, method_name)
+      return Method(method_value, arguments, defaults)
 
-    Args:
-      fn: user DoFn to invoke
-      args: positional side input arguments (static and placeholder), if any
-      kwargs: keyword side input arguments (static and placeholder), if any
-      side_inputs: list of sideinput.SideInputMaps for deferred side inputs
-      windowing: windowing properties of the output PCollection(s)
-      context: a DoFnContext to use (deprecated)
-      tagged_receivers: a dict of tag name to Receiver objects
-      logger: a logging module (deprecated)
-      step_name: the name of this step
-      logging_context: a LoggingContext object
-      state: handle for accessing DoFn state
-      scoped_metrics_container: Context switcher for metrics container
-    """
-    self.step_name = step_name
-    self.window_fn = windowing.windowfn
-    self.tagged_receivers = tagged_receivers
-    self.scoped_metrics_container = (scoped_metrics_container
-                                     or ScopedMetricsContainer())
+    signature.do_fn = do_fn
+    signature.process_method = _create_do_fn_method(do_fn, 'process')
+    signature.start_bundle_method = _create_do_fn_method(do_fn, 'start_bundle')
+    signature.finish_bundle_method = _create_do_fn_method(
+        do_fn, 'finish_bundle')
 
-    global_window = GlobalWindow()
+    return signature
 
-    # Need to support multiple iterations.
-    side_inputs = list(side_inputs)
 
-    if logging_context:
-      self.logging_context = logging_context
-    else:
-      self.logging_context = get_logging_context(logger, step_name=step_name)
+class DoFnInvoker(object):
+  """An abstraction that can be used to execute DoFn methods.
 
-    # Optimize for the common case.
-    self.main_receivers = as_receiver(tagged_receivers[None])
+  A DoFnInvoker describes a particular way for invoking methods of a DoFn
+  represented by a given DoFnSignature."""
 
-    # TODO(sourabh): Deprecate the use of context
-    if state:
-      assert context is None
-      self.context = DoFnContext(self.step_name, state=state)
-    else:
-      assert context is not None
-      self.context = context
+  def __init__(self, signature):
+    self.signature = signature
 
-    class ArgPlaceholder(object):
-      def __init__(self, placeholder):
-        self.placeholder = placeholder
+  @staticmethod
+  def create_invoker(
+      signature, use_simple_invoker, context, side_inputs, input_args,
+      input_kwargs):
+    if use_simple_invoker:
```

Comment:
Docstring? 

Also, it feels like "use_simple_invoker" (and perhaps other arguments) should be deduced here, not passed in.